src/Rector/ClassMethod/ResponseReturnTypeControllerActionRector.php

```diff
@@ -98,17 +98,88 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
-        // has redirect return response
-        if ($this->hasRedirectReturnResponse($node)) {
+        if ($this->isRedirectResponseReturn($node)) {
             $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\RedirectResponse');
-        } else {
-            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\Response');
+
+            return $node;
+        }
+        if ($this->isBinaryFileResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\BinaryFileResponse');
+
+            return $node;
+        }
+        if ($this->isJsonResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\JsonResponse');
+
+            return $node;
+        }
+        if ($this->isStreamedResponseReturn($node)) {
+            $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\StreamedResponse');
+
+            return $node;
         }
 
+        $node->returnType = new FullyQualified('Symfony\Component\HttpFoundation\Response');
+
         return $node;
     }
 
-    private function hasRedirectReturnResponse(ClassMethod $classMethod): bool
+    private function isRedirectResponseReturn(ClassMethod $classMethod): bool
+    {
+        $returns = $this->betterNodeFinder->findInstanceOf($classMethod, Return_::class);
```

Comment:
You need to use `findInstancesOfInFunctionLikeScoped($classMethod, Return_::class)` to ensure it not check to inner function/class/closure.

Reply:
I mean, to avoid something like this:

```
public function foo()
{
      function () {
             // this inner return should not be marked as found Node.
             return $this->foo();
      };
}
```